### PR TITLE
[MM-?] Patches for deficiencies of the current security Audit system

### PR DESCRIPTION
**Pull request by giuliocomi**
This pull request contains some fixes to the current Audit system.
The details of the weaknesses have been provided in [issue #13256]("https://github.com/mattermost/mattermost-server/issues/13256").
Detailed document available here [enhanced_audit_system.pdf](https://github.com/mattermost/mattermost-server/files/3906421/enhanced_audit_system.pdf).

Adding audit logging of following events:
• Account lockdown
• API rate limiting threshold reached
• Log which was the past roles upon granting user the new roles
• Change of email address
• Incorrect anti-CSRF token submitted
• Insufficient permission errors
• Improved the identification of the action via the mapping of HTTP method + endpoint URL

Many thanks to @DSchalla for the valueable feedbacks.

### CODE REVIEW

**Comment by DSchalla** [comment]
Thanks for your PR @giuliocomi :tada: I edited the Ticket ID out of the header and will create one tomorrow and add it here. I fear the pull request has a problem with gofmt not passing, would you like to take a look? Otherwise happy to help.

**Comment by jasonblais** [comment]
@DSchalla @crspeller @wiersgallak friendly reminder to help with reviews :)

File: web/context.go [hunk 1]
```diff
@@ -24,19 +24,24 @@ type Context struct {
...
+	audit := &model.Audit{UserId: c.App.Session.UserId, IpAddress: c.App.IpAddress, Action: action, ExtraInfo: extraInfo, SessionId: c.App.Session.Id}
+
 	if err := c.App.Srv.Store.Audit().Save(audit); err != nil {
 		c.LogError(err)
 	}
 }
 
 func (c *Context) LogAuditWithUserId(userId, extraInfo string) {
+	method := []string{c.App.HttpMethod, " "}
+	action := strings.Join(method, c.App.Path)
```
Comment (crspeller) [inline web/context.go]: Maybe use `fmt.Sprintf`?

**Comment by hanzei** [comment]
@DSchalla Should this PR still go into 5.19?

**Comment by teresa-novoa** [comment]
@DSchalla friendly reminder to review this PR, thanks!

**Comment by DSchalla** [comment]
With https://github.com/mattermost/mattermost-server/pull/13967 being merged a few weeks ago I fear I need to close this PR, @giuliocomi very much appreciate the contribution and sorry we dropped the ball here and didn't follow up earlier. Would love to see your thoughts about the newly implemented auditing system!

**Comment by teresa-novoa** [comment]
Thanks @DSchalla!

**Pull request closed without merging**